### PR TITLE
Add parallelism to GCE cluster upgrade

### PR DESCRIPTION
**Pull request by shyamjvs**
Fixes https://github.com/kubernetes/kubernetes/issues/48373
Should allow upgrading 500-node cluster (1.6->1.7) in < 1 hr. It currently takes ~1.5 day.
Though it is the duty of the upgrader to choose the right parallelism in order to avoid disrupting too many pods.

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews @kubernetes/sig-scalability-misc @mikedanese @gmarek

### CODE REVIEW

**Comment by shyamjvs** [comment]
@wojtek-t Do you think this is reasonable? Might not be much useful for production clusters, but for testing if large cluster upgrades work this might help(?)

**Comment by shyamjvs** [comment]
friendly ping @wojtek-t @maisem

**Comment by wojtek-t** [comment]
I think exposing parallelism as an argument is useful.
That said, even for testing purposes I wouldn't make it big (e.g. for 1000-node cluster, it's probably fine to set it to 10, but not to 100).

**Comment by wojtek-t** [comment]
I didn't review the PR though. If you want me to review it, please reassign it to me.

**Comment by shyamjvs** [comment]
/assign @wojtek-t 
Thanks.

**Comment by shyamjvs** [comment]
> That said, even for testing purposes I wouldn't make it big (e.g. for 1000-node cluster, it's probably fine to set it to 10, but not to 100).

I totally agree.

**Comment by shyamjvs** [comment]
@wojtek-t friendly ping

File: cluster/gce/upgrade.sh [hunk 1]
```diff
@@ -39,6 +39,7 @@ function usage() {
   echo "  -M:  Upgrade master only"
   echo "  -N:  Upgrade nodes only"
   echo "  -P:  Node upgrade prerequisites only (create a new instance template)"
+  echo "  -c:  Upgrade NODE_UPGRADE_PARALLELISM nodes in parallel (default=1) within a single instance group. The IGs themselves are dealt serially."
```
Comment (wojtek-t) [inline cluster/gce/upgrade.sh]: s/IGs/MIGs/

File: cluster/gce/upgrade.sh [hunk 2]
```diff
@@ -362,95 +458,30 @@ function do-node-upgrade() {
...
-        cordoned=$("${KUBE_ROOT}/cluster/kubectl.sh" get node "${instance}" --output='jsonpath={.status.conditions[?(@.type == "SchedulingDisabled")].status}')
-        ready=$("${KUBE_ROOT}/cluster/kubectl.sh" get node "${instance}" --output='jsonpath={.status.conditions[?(@.type == "Ready")].status}')
-        if [[ "${cordoned}" == 'True' ]]; then
-          echo "Node ${instance} is still not ready: SchedulingDisabled=${ready}"
-        elif [[ "${ready}" != 'True' ]]; then
-          echo "Node ${instance} is still not ready: Ready=${ready}"
-        else
-          echo "Node ${instance} Ready=${ready}"
-          break
+    proc=${node_upgrade_parallelism}
```
Comment (wojtek-t) [inline cluster/gce/upgrade.sh]: What is "proc" ?

Comment (wojtek-t) [inline cluster/gce/upgrade.sh]: If this is some shortcut, can we avoid it?

File: cluster/gce/upgrade.sh (hunk 1 above)
Comment (shyamjvs) [inline cluster/gce/upgrade.sh]: Done.

File: cluster/gce/upgrade.sh (hunk 2 above)
Comment (shyamjvs) [inline cluster/gce/upgrade.sh]: Just a counter for no. of more processes that can be started in parallel. Renamed it to `process_count_left`.

**Comment by shyamjvs** [comment]
Squashed commits and reapplying lgtm based on https://github.com/kubernetes/kubernetes/pull/48429#issuecomment-319041089.

**Pull request merged**